eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
       if last.pow > msg.pow or
         (last.pow == msg.pow and last.env.expiry > msg.env.expiry):
         # The new message has less pow or will expire earlier - drop it
-        self.items.del(self.items.len() - 1)
+        return false
 
-  self.items.insert(msg, self.items.lowerBound(msg, cmpPow))
+      self.items.del(self.items.len() - 1)
+      self.itemHashes.excl(last)
+
+  # check for duplicate
+  # NOTE: Could also track if duplicates come from the same peer and disconnect
+  # from that peer. Is this tracking overhead worth it though?
```

Comment:
There is a notion of Peer reputation in some of the other clients.

I guess it's enough for us to keep general stats for each peer (how often we get duplicates from that peer) and if a certain threshold is reached, we may disconnect and refuse new connections (up to some time perhaps). But all of this is more in the "nice to have" category rather than a top priority item.

Reply:
Yes, what I mean here is that some stats for the the peer reputation are more easy to track than others in terms of overhead (and complexity). 

E.g. receiving `invalid` or `not allowed` envelopes. It is a matter of adding counters per peer. The check is already done anyhow.

But for the duplicates, we would have to add for each message which peer has send it to us already. And check it when receiving the message again of course. As duplicates are bound to happen anyhow, and you can not punish a peer if it is their 'first' duplicate of a specific message. But OK, maybe this is acceptable as the amount of peers is not huge (and should also be configurable)